Lesson17

JuicyFlash opened the pull request:
Приложение развёрнуто на Heroku
Изменено название приложения
Создан аккаунт администратора
Добавлена страничка с формой обратной связи
Заполнено данными (seed)

Учётка админа:
tankedet@bk.ru
123456

https://nikolaev-guru.herokuapp.com/

JuicyFlash committed:
Переход на PostGresSQL

JuicyFlash committed:
Не создавался новый пользователь - поправил

JuicyFlash committed:
ENV только для test и dev

JuicyFlash committed:
Настройка почты

JuicyFlash committed:
Настройки почты изменены на mail

JuicyFlash committed:
Настройки почты изменены на mail

JuicyFlash committed:
Настройки почты изменены на gmail

JuicyFlash committed:
Настройки почты изменены на gmail

JuicyFlash committed:
Настройки почты sendmail

JuicyFlash committed:
Настройки почты gmail (рабочий вариант)

JuicyFlash committed:
Обратная связь

JuicyFlash committed:
Фикс багов:
начать тест без вопросов
перейти к след вопросу без выбора ответа
при создании гиста
при удалении теста, который кто-то проходил

BubuntuClu commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -38,11 +38,21 @@ def update_inline
   end
 
   def start
```
в админке не нужен старт

BubuntuClu commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -38,11 +38,21 @@ def update_inline
   end
 
   def destroy
+    @test.questions.each do | qs |
```
это костыльное решение. у рельсы есть функционал на удаление связанных ассоциаций. опиция `dependent` для ассоциаций, посмотри в эту сторону

BubuntuClu replied:
так, я долистал вниз. эти опции у тебя прописаны, зачем тогда руками удалять это все?

BubuntuClu commented on app/controllers/processed_tests_controller.rb:
```diff
@@ -37,7 +41,8 @@ def gist
 
   def set_processed_test
-    @processed_test = ProcessedTest.find(params[:id])
+    @processed_test = ProcessedTest.find!(params[:id])
```
почему тут используешь `find!` метод?

BubuntuClu replied:
а, увидел, это реализация кастомная. а зачем тебе обработчик выносить в отдельный метод модели, а не как в других контроллерах сделать обработчик в контроллере?

BubuntuClu commented on app/controllers/tests_controller.rb:
```diff
@@ -7,8 +7,12 @@ def index
   def start
-    current_user.tests.push(@test)
-    redirect_to current_user.processed_test(@test)
+    if @test.questions.count > 0
```
поставь себя на место пользователя. ты зашел на сайт, видишь доступный тест, хочешь его пройти, а тебе говорят "нельзя проходить, нет вопросов". что ты подумаешь в этот момент?

решить проблему с отсутствием вопросов надо другим способом.

да и если будет один вопрос, но не будет ответов у этого вопроса, думаешь это хорошая ситуация?

BubuntuClu commented on app/models/question.rb:
```diff
@@ -2,8 +2,8 @@
 
-  has_many :answers
-  has_many :gists
+  has_many :answers, dependent: :delete_all
```
почему `delete_all` а не `destroy_all` или может быть даже `destroy`?

JuicyFlash committed:
Убран костыль с удалением (не полностью, оставлен processed_test, как убрать непонятно - ,без него ругается на ссылку questions->processed_tests при удалении теста)
Убран кастомный метод find!
Переработан случай когда у теста нет вопросов - добавлен атрибут теста (готовность), устанавливаемый админом, не готовые тесты пользователь не видит, значение атрибута выведено в список тестов админа, редактировать можно на форме редактирования теста

JuicyFlash committed:
Fix migration

BubuntuClu commented on app/views/tests/index.html.erb:
```diff
@@ -7,6 +7,7 @@
 </h1>
 <div class="col-7 items-list">
 <% @tests.each do |test| %>
+<% if test.ready %>
```
представь, у тебя 1000000 тестов в системе, большинство в процессе формирования к показу. из этих 1000000 у тебя готовых только 7. стоит ли доставать все 1000000 и на каждый делать проверку готов ли он? может можно как-то облегчить задачу?

BubuntuClu replied:
это условие еще нужно? сейчас у тебя все тесты готовые

BubuntuClu commented on db/migrate/20220823203044_add_ready_to_tests.rb:
```diff
@@ -0,0 +1,5 @@
+class AddReadyToTest < ActiveRecord::Migration[6.1]
+  def change
+    add_column :tests, :ready, :bool, default: 0
```
если bool, то почему деолтное значение 0,а не false?

JuicyFlash committed:
Новый Scope метод Test.ready, список тестов пользователя формиркется с его использованием

Правка в миграцию

JuicyFlash committed:
Лишнее условие проверки tests index
